Move -Werror from packages into project

locallycompact opened the pull request:
We should not have Werror on by default as it creates friction for downstream users consuming published releases of our libraries. If somebody includes our library in a build plan which does not match our development exactly, then that can issue a warning that will throw an error if Werror is set in our cabal files. Things which are innocuous such as StarIsType deprecations or ~ requires TypeOperator deprecations.

Users would then have to override our package in their build plan to add the development flag, which is needless work for developers.

locallycompact committed:
Remove Werror and move it to CI only

ch1bo commented on nix/hydra/project.nix:
```diff
@@ -53,33 +53,49 @@ let
-      }
+        strict-containers.doHaddock = false;
+
+        # -Werror for CI
```
This is not only enabling Werror for CI but for any `nix build` involving these packages. Either we put `-Werror` somewhere else or update this comment.

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -252,9 +244,7 @@ benchmark tx-cost
-  if flag(hydra-development)
-    -- NOTE(SN): should fix HLS choking on PlutusTx plugin
-    ghc-options: -fplugin-opt PlutusTx.Plugin:defer-errors
+  ghc-options: -fplugin-opt PlutusTx.Plugin:defer-errors
```
Instead of turning them on for the full package, we should instead reach to module-level disabling of this via `{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:defer-errors #-}` pragmas on the affected modules.

ch1bo replied:
Doing this removed the need to do quite some haddock workarounds: a75c0f6765f42015ba9f84a0aea06297feb78daa

ch1bo committed:
Move PlutusTx.Plugin:defer-errors into modules

This is only needed for HLS to not choke on the plutus-tx plugin. As we
recently removed the hydra-development flag and enabled it on package
level, we moved this further "down" into each affected module and this
makes Haddock pick it up correctly.

ch1bo reviewed (approved):
I think this is fine and will make all our `nix build` invocations fail with warnings as errors, making CI on PRs report all warnings, but also give developers an easy way to run a "Werror-build" by doing `nix build` (assuming nobody used the nix invocation for development)

I want a review of @abailly-iohk on this as well as he had an opinion on disabling `Werror` in the past.

locallycompact commented:
Here are complaints of this in other published libraries that would also be true for us when we publish to CHaP.

https://github.com/input-output-hk/cardano-base/issues/372

ch1bo commented:
@abailly-iohk @locallycompact in the discussion of https://github.com/input-output-hk/cardano-base/issues/372 there is a nice solution:

Turn `-Werror` on in `cabal.project`.

That way, any `cabal` build using our repository will treat warnings as errors (including CI), but downstream projects depending on the `.cabal` files will not be impacted and developers can turn it off (as mentioned in the issue above).

I think I would prefer this option and will request it as a change.

ch1bo reviewed (changes requested):
As discussed, turning `-Werror` on in `cabal.project` is more in-line with what we have today. Furthermore it does keep the nix build more similar to a non-nix build.

We could also put a line of advice into `CONTRIBUTING.md` how to turn it off (e.g. what to put in a `cabal.project.local`)

ch1bo committed:
Move -Werror to cabal.project

This enables warnings as errors for all local packages and thus fails
builds inside the nix toolchain, but also non-nix invocations of cabal.

ch1bo committed:
Note about turnign off -Werror in contribution guidelines

ch1bo reviewed (approved):
@abailly-iohk Changed it to be configured in the `cabal.project`. Do you approve?
